pkg/cmd/roachtest/tests/cluster_to_cluster.go

```diff
@@ -314,82 +394,169 @@ func setupC2C(
+			sp.t.L().Printf("%s: %.2f", name, metricSnap[name])
+		}
+		return metricSnap
+	}
 }
 
-type replicateKV struct {
-	readPercent int
+func (sp *replicationTestSpec) preStreamingWorkload(ctx context.Context) {
+	if initCmd := sp.workload.sourceInitCmd(sp.setup.src.name, sp.setup.src.nodes); initCmd != "" {
+		sp.t.Status("populating source cluster before replication")
+		initStart := timeutil.Now()
+		sp.c.Run(ctx, sp.setup.workloadNode, initCmd)
+		sp.t.L().Printf("src cluster workload initialization took %s minutes",
+			timeutil.Since(initStart).Minutes())
+	}
 }
 
-func (kv replicateKV) sourceInitCmd(tenantName string, nodes option.NodeListOption) string {
-	return ""
+func (sp *replicationTestSpec) startReplicationStream() int {
+	streamReplStmt := fmt.Sprintf("CREATE TENANT %q FROM REPLICATION OF %q ON '%s'",
+		sp.setup.dst.name, sp.setup.src.name, sp.setup.src.pgURL)
+	sp.setup.dst.sysSQL.Exec(sp.t, streamReplStmt)
+	return getIngestionJobID(sp.t, sp.setup.dst.sysSQL, sp.setup.dst.name)
 }
 
-func (kv replicateKV) sourceRunCmd(tenantName string, nodes option.NodeListOption) string {
-	// added --tolerate-errors flags to prevent test from flaking due to a transaction retry error
-	return fmt.Sprintf(`./workload run kv --tolerate-errors --init --read-percent %d {pgurl%s:%s}`,
-		kv.readPercent,
-		nodes,
-		tenantName)
+func (sp *replicationTestSpec) runWorkload(ctx context.Context) error {
+	return sp.workload.runDriver(ctx, sp.c, sp.t, sp.setup)
 }
 
-type replicationTestSpec struct {
-	// name specifies the name of the roachtest
-	name string
-
-	// srcodes is the number of nodes on the source cluster.
-	srcNodes int
+func (sp *replicationTestSpec) waitForHighWatermark(ingestionJobID int, wait time.Duration) {
+	testutils.SucceedsWithin(sp.t, func() error {
+		info, err := getStreamIngestionJobInfo(sp.setup.dst.db, ingestionJobID)
+		if err != nil {
+			return err
+		}
+		if info.GetHighWater().IsZero() {
+			return errors.New("no high watermark")
+		}
+		return nil
+	}, wait)
+}
 
-	// dstNodes is the number of nodes on the destination cluster.
-	dstNodes int
+func (sp *replicationTestSpec) getWorkloadTimeout() time.Duration {
+	if sp.additionalDuration != 0 {
+		return sp.additionalDuration
+	}
+	return sp.timeout
+}
 
-	// cpus is the per node cpu count.
-	cpus int
+// getReplicationRetainedTime returns the `retained_time` of the replication
+// job.
+func (sp *replicationTestSpec) getReplicationRetainedTime() time.Time {
+	var retainedTime time.Time
+	sp.setup.dst.sysSQL.QueryRow(sp.t,
+		`SELECT retained_time FROM [SHOW TENANT $1 WITH REPLICATION STATUS]`,
+		roachpb.TenantName(sp.setup.dst.name)).Scan(&retainedTime)
+	return retainedTime
+}
 
-	// pdSize specifies the pd-ssd volume (in GB). If set to 0, local ssds are used.
-	pdSize int
+func (sp *replicationTestSpec) stopReplicationStream(ingestionJob int, cutoverTime time.Time) {
+	sp.setup.dst.sysSQL.Exec(sp.t, `SELECT crdb_internal.complete_stream_ingestion_job($1, $2)`,
```

Comment:
While we are here, we probably want to update this to use `ALTER TENANT ... COMPLETE REPLICATION TO`

Reply:
Done.